File: silumawig.R
silumawig = function(X,Y){

  
  # Initialize
  X = X[order(-X)]
  Y = Y[order(-Y)]  
  d = 0
  N = min(length(X),length(Y))
  X = X[1:N]
  Y = Y[1:N]
  Z = abs(X-Y)
  l = which(Z==max(Z))[1]
  d=0
  

  if(N==1){
    d = abs(X[1]-Y[1])/2
    return(d)
  }
  
  if(Y[l]>X[l]){
    S=Y
    Y=X
    X=S
  }
  
  if(length(which(X>=Y))==N){
    d = abs(X[l]-Y[l])/2
    return(d)
  }
  else{
    m=which((X-Y)==min(X-Y))[1]
   if(Y[m]-X[m] > Y[l] & m<l){
     return(silumawig(X[1:l],Y[1:l]))
   }
    else if(X[l]-Y[l] > X[m] & m>l){
      return(silumawig(X[1:m],Y[1:m]))
    }
    else{
      d = (X[l]-Y[l]-X[m]+Y[m])/2
      return(d)
    }
  }
}
  

File: silumawig.py
'''Compute the dimension 0 shifted bottleneck distance between two persistence 
    diagrams X and Y.
'''

import numpy as np




def silumawig(X,Y):
    X=-np.sort(-X)
    Y=-np.sort(-Y)
    d=0
    shift=0
    
    N = min(len(X),len(Y))
    if N==1:
        d=abs(X[0]-Y[0])/2
        shift=d
        return d,shift
    X=X[0:N]
    Y=Y[0:N]
    Z = abs(X-Y)
    l = np.argmax(Z)



    if Y[l]>=X[l]:
        X_copy=X
        X=Y
        Y=X_copy
    if len(np.where(X>=Y)[0])==N:
        d=(X[l]-Y[l])/2
        shift=d
        return d,shift
            
    else: 
        m=np.argmin(X-Y)
        if Y[m] - X[m] > Y[l] and m<l: 
            return silumawig(X[0:l],Y[0:l])
        elif X[l] - Y[l] > X[m] and m>l:
            return silumawig(X[0:m],Y[0:m])
        else:
            d=(X[l]-Y[l]-X[m]+Y[m])/2
            shift=(X[l]-Y[l]+X[m]-Y[m])/2
            return d,shift           

    
